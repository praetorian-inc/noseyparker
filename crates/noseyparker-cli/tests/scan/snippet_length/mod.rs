use super::*;

#[test]
fn scan_invalid_snippet_length() {
    let scan_env = ScanEnv::new();
    let input = scan_env.input_dir("empty_dir");
    noseyparker_failure!(
        "scan",
        "--datastore",
        scan_env.dspath(),
        input.path(),
        "--snippet-length=-1"
    )
    .stderr(is_match("error: invalid value .* for .*: invalid digit found in string"));
}

#[test]
fn scan_changing_snippet_length() {
    let scan_env = ScanEnv::new();
    let input = scan_env.large_input_file_with_secret("input.txt");

    // first scan with non-default short snippet length
    noseyparker_success!("scan", "-d", scan_env.dspath(), input.path(), "--snippet-length=16")
        .stdout(match_scan_stats("1.41 KiB", 1, 1, 1));

    assert_cmd_snapshot!(noseyparker_success!("summarize", "-d", scan_env.dspath()));

    with_settings!({
        filters => get_report_stdout_filters(),
    }, {
        assert_cmd_snapshot!(noseyparker_success!("report", "-d", scan_env.dspath()));
    });

    let cmd = noseyparker_success!("report", "-d", scan_env.dspath(), "--format=json");
    let json_output: serde_json::Value = serde_json::from_slice(&cmd.get_output().stdout).unwrap();
    with_settings!({
        redactions => get_report_json_redactions(),
    }, {
        assert_json_snapshot!(json_output);
    });

    // now rescan with longer snippet length
    noseyparker_success!("scan", "-d", scan_env.dspath(), input.path(), "--snippet-length=32")
        .stdout(match_scan_stats("1.41 KiB", 1, 0, 1));

    assert_cmd_snapshot!(noseyparker_success!("summarize", "-d", scan_env.dspath()));

    with_settings!({
        filters => get_report_stdout_filters(),
    }, {
        assert_cmd_snapshot!(noseyparker_success!("report", "-d", scan_env.dspath()));
    });

    let cmd = noseyparker_success!("report", "-d", scan_env.dspath(), "--format=json");
    let json_output: serde_json::Value = serde_json::from_slice(&cmd.get_output().stdout).unwrap();
    with_settings!({
        redactions => get_report_json_redactions(),
    }, {
        assert_json_snapshot!(json_output);
    });
}
