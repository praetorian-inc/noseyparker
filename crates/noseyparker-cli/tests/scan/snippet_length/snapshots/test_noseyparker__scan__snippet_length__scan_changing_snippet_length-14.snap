---
source: crates/noseyparker-cli/tests/scan/snippet_length/mod.rs
expression: json_output
---
[
  {
    "comment": null,
    "finding_id": "d551329ba5578559646aa49467be47e9d496578d",
    "groups": [
      "Z2hwX1hJeEI3S01OZEFyM3pxV3RRcWhFOTRxZ2xIcU96bjFEMXN0Zw=="
    ],
    "matches": [
      {
        "blob_id": "96add51d102b68a8eb16cd104f6a79edc27bb61d",
        "blob_metadata": {
          "charset": null,
          "id": "96add51d102b68a8eb16cd104f6a79edc27bb61d",
          "mime_essence": "text/plain",
          "num_bytes": 1448
        },
        "comment": null,
        "groups": [
          "Z2hwX1hJeEI3S01OZEFyM3pxV3RRcWhFOTRxZ2xIcU96bjFEMXN0Zw=="
        ],
        "location": {
          "offset_span": {
            "end": 775,
            "start": 735
          },
          "source_span": {
            "end": {
              "column": 51,
              "line": 30
            },
            "start": {
              "column": 12,
              "line": 30
            }
          }
        },
        "provenance": [
          {
            "kind": "file",
            "path": "<ROOT>/input.txt"
          }
        ],
        "redundant_to": [],
        "rule_name": "GitHub Personal Access Token",
        "rule_structural_id": "f6c4fca24a1c7f275d51d2718a1585ca6e4ae664",
        "rule_text_id": "np.github.1",
        "score": null,
        "snippet": {
          "after": "\n\nfunction lorem(ipsum, dolor = ",
          "before": "ta\nUSERNAME=the_dude\nGITHUB_KEY=",
          "matching": "ghp_XIxB7KMNdAr3zqWtQqhE94qglHqOzn1D1stg"
        },
        "status": null,
        "structural_id": "02f264f3a42f38d96d0069e4b91e3d3e66bf8b08"
      }
    ],
    "mean_score": null,
    "num_matches": 1,
    "num_redundant_matches": 0,
    "rule_name": "GitHub Personal Access Token",
    "rule_structural_id": "f6c4fca24a1c7f275d51d2718a1585ca6e4ae664",
    "rule_text_id": "np.github.1",
    "statuses": []
  }
]
